### ISSUE TITLE
Fix for POKE and VERA

### ISSUE THREAD

**Issue by tomxp411** [commanderx16/x16-rom#65]
According to the 65C02 data sheet on page 30, the 65C02 processor performs two memory accesses during an STA (zp),Y. (Make sure to reference note 1, which stipulates 1 more cycle count on write.)

This breaks the POKE command when used on VERA, because it will put data in the wrong location on screen. When used with any non-zero increment value, the POKEd character will be one step to the right of where intended, and the memory pointer will advance two locations. 

The fix is very simple. The POKE command is in code17.s line 76:
```
poke jsr getnum
	 txa
	 ldy #0
	 sta (poker),y
	 rts
```
This should change to 
```
poke  jsr getnum
	  txa
	  sta (poker)
	  rts
```
This change does break 6502 compatibility, since STA (zp) without indexing is a new opcode.

**Comment by mist64** [comment]
The data sheet in question is the version from March 2000:
http://archive.6502.org/datasheets/wdc_w65c02s_mar_2000.pdf

**Comment by mist64** [comment]
I tried to research this further, but couldn't find any information on what the 65C02 does in the extra cycle. MAME, for example doesn't seem to account for it: https://github.com/mamedev/mame/blob/ee1e4f9683a4953cb9d88f9256017fcbc38e3144/src/devices/cpu/m6502/om65c02.lst#L680

**Comment by akumanatt** [comment]
Indexed store always spend one extra internal cycle, page crossing or not. And the extra read comes from that internal cycle, which will read from `(addr & $ff00) | ((addr + Y) & $00ff)`. And this was NMOS 6502 behavior which W65C02S claimed to fixed it in its datasheet: it will read the last instruction byte instead. However in W65C816S's datasheet, emulation mode still does this invalid read. So now I'm not sure about this conflicting information...

**Comment by akumanatt** [comment]
So we (I, mist64 and Wavicle) did a test on a real 65C02 and can confirm that `sta (zp), y` takes 6 cycles, reads zp+1 twice instead of invalid address and only writes to the destination once.

**Comment by akumanatt** [comment]
Alright, so I and Wavicle tested this on a real W65C02S and can confirm that `sta (zp), y` always take 6 cycles regardless of page crossing. And it reads zp+1 twice on an extra cycle instead of NMOS 6502 behavior.

However, this fix is still useful since `sta (zp)` always take 5 cycles, which is one less. And NMOS 6502 code needs to have a different code to handle this anyway.

**Comment by mist64** [comment]
I don't understand this part:

> And NMOS 6502 code needs to have a different code to handle this anyway.

**Comment by akumanatt** [comment]
It means that for that CPU, you can't neither have `sta (zp)` nor `sta (zp),y`, as it will write to (zp) twice. Since I remember you still want NMOS 6502 compatibility in this. The fix would for it would involve writing `sta dst` code to the RAM and execute it.

**Comment by mist64** [comment]
Yes, I'd like to BASIC codebase to remain NMOS 6502 compatible. Yes, this code is incorrect on the NMOS 6502 if the hardware being written to has side effects when writing to it twice.

It's not a problem on the X16 or the C64 though. The former has the CPU fixed, the latter doesn't have any hardware with such side effects.

I'll keep it open, but low-prio.

**Comment by mist64** [comment]
I guess the best way to fix this is to write `8D lo hi 60` into RAM and executing it. BASIC should have some continguous bytes in either `ZPKERNAL` or `KVAR` that can be reused temporarily.
